### ISSUE TITLE
Extreme Memory Usage with LoadListDataAsStreamAsync

### ISSUE THREAD

**Issue by DaleyKD** [pnp/pnpcore#545]
## Category

- [x] Bug

## Describe the bug
When using `list.LoadListDataAsStreamAsync` in combination with a LOT of list items, our K8S pod runs out of memory.

## Steps to reproduce
1. Generate a lot of files.
``` ps1
for ($i = 1; $i -lt 2500; $i++) {
	$fname = "File_" + $i + ".tst"
	fsutil file createnew $fname (Get-Random -Minimum 25 -Maximum 10000)
}
```
2. Upload these files to a Document library.
3. Run this code:
``` csharp
private static readonly PnPContextOptions ContextOptions = new PnPContextOptions()
{
    AdditionalSitePropertiesOnCreate = new List<Expression<Func<ISite, object>>>()
    {
        s => s.Url
    },
    AdditionalWebPropertiesOnCreate = new List<Expression<Func<IWeb, object>>>()
    {
        w => w.ServerRelativeUrl,
        w => w.Title,
        w => w.AppInstanceId
    }
};

private readonly int _queryRowLimit = 1000;

public async Task ProcessWebWithCamlQuery()
{
    var myUri = new Uri("https://mysharepoint.sharepoint.com/sites/site-test");
    using (var context = await _pnpContextFactory.CreateAsync(myUri, ContextOptions))
    {
        var lists = GetListsToProcess(context.Web);
        var viewXml = GetCamlQuery();

        foreach (var list in lists)
        {
            var paging = true;
            string nextPage = null;
            while (paging)
            {
                var output = await list.LoadListDataAsStreamAsync(new RenderListDataOptions()
                {
                    ViewXml = viewXml,
                    RenderOptions = RenderListDataOptionsFlags.ListData,
                    Paging = nextPage
                }).ConfigureAwait(false);

                if (output.ContainsKey("NextHref") && output["NextHref"] != null)
                {
                    nextPage = output["NextHref"].ToString()?[1..];
                }
                else
                {
                    paging = false;
                }
            }

            var listItems = list.Items.AsRequested();
            Console.WriteLine(GC.GetTotalMemory(false));
            Console.WriteLine("Process the listItems here");
        }
    }

    Console.WriteLine("Done processing list");
    Console.WriteLine(GC.GetTotalMemory(false));
}

private string GetCamlQuery()
{
    return $@"<View Scope='RecursiveAll'>
            <RowLimit Paged='TRUE'>{_queryRowLimit}</RowLimit>
            <Query>
              <Where>
                <Or>
                    <BeginsWith>
                      <FieldRef Name='ContentTypeId' />
                      <Value Type='ContentTypeId'>0x0101</Value>
                    </BeginsWith>
                    <BeginsWith>
                      <FieldRef Name='ContentTypeId' />
                      <Value Type='ContentTypeId'>0x0120</Value>
                    </BeginsWith>
                </Or>
              </Where>
            </Query>
        </View>";
}

private static readonly Expression<Func<IList, bool>> ListPredicate = l =>
    l.NoCrawl == false &&
    l.Hidden == false &&
    l.IsApplicationList == false &&
    l.TemplateType == ListTemplateType.DocumentLibrary &&
    l.ListItemEntityTypeFullName != "SP.Data.AppPackagesListItem" &&
    l.ListItemEntityTypeFullName != "SP.Data.FormServerTemplatesItem";

private static readonly Expression<Func<IList, object>>[] ListProperties = {
    l => l.Title,
    l => l.RootFolder.QueryProperties(f => f.ServerRelativeUrl),
    l => l.ContentTypes.QueryProperties(
        ct => ct.Name,
        ct => ct.Fields
            .QueryProperties(
                f => f.InternalName,
                f => f.Title,
                f => f.Hidden,
                f => f.FieldTypeKind))
};

private static IEnumerable<IList> GetListsToProcess(IWeb web)
{
    // The Id (whatever is the Key) is populated automatically; no need to explicitly add it here
    return web.Lists.Where(ListPredicate).QueryProperties(ListProperties);
}
```

## Expected behavior
I realize that we're pumping a lot of data into memory.  However, I haven't had a chance to dig into `ListItem` to determine what is holding on to SO much data that memory just keeps skyrocketing with each 1000 items.

Ideally, we don't skyrocket in memory.  Even if we try `myListOfListItems.AddRange(list.Items.AsRequested()); list.Items.Clear();`, then `myListOfListItems` is still holding on to a lot of memory.

## Environment details (*development & target environment*)
- **SDK version**: 1.3.0
- **OS**: Windows 10
- **SDK used in**: ASP.Net Web app
- **Framework**: .NET Core 5
- **Tooling**: Visual Studio 2019
- **Additional details**: *The more context you can provide, the easier it is (and therefore quicker) to help.*

### Additional context
Some customers most. definitely. exceed the LVT.  To try and make the rest of our code "efficient," we were trying to load up the all list items first, then use them in parallel later.

We've tried to use a new `PnPContext` per 1000 items, but you and I both know that's an expensive call to keep doing (don't want to keep making the same Site/Web/List API endpoint calls over and over).

I was hoping that maybe bringing this up may cause an immediate lightbulb to go off and you'd know why/how to fix.

Thanks, Bert, for everything.

**Comment by jansenbe** [comment]
@DaleyKD : memory consumption is indeed quite high, did implement a test case and got below. Will further dig in tomorrow:

![image](https://user-images.githubusercontent.com/7451219/133320633-05105d28-25b0-4d0b-bbf1-d2db2259468b.png)

**Comment by DaleyKD** [comment]
@jansenbe : GREATLY appreciated!  I suspect it's potentially the number of items in the TransientDictionary, but I'm not sure. _(I haven't been able to be on this task for my team; this was found by a teammate!)_

**Comment by jansenbe** [comment]
@DaleyKD : I've been doing some more testing and as expected the memory consumption is driven by loading more list items. A list item stores it's data in a `Dictionary<string, object>` and it's metadata in a `Dictionary<string, string>` and for 2500 list items having a ton of fields this adds up. What you can do is first of all consider only loading the fields you really need by using the `<ViewFields>` option in your CAML query, this will not only lower the amount of memory needed but also will speed up things due to lesser processing needed. 

Given the used method (`LoadListDataAsStreamAsync`) is a so called `Load` method the gathered data is kept in memory, `Get` methods return data as a variable scoped to the current block, so going out of scope automatically. For data loaded in the domain model you can do something similar by clearing the list items (or full list). In my testing I used `list.Items.Clear()` and that resulted in dropping the list item objects from the managed heap:

![image](https://user-images.githubusercontent.com/7451219/133398472-14692dbf-8e86-4b64-a2f8-c97b1286b94f.png)

When checking memory consumption using `GC.GetTotalMemory(false)` you might not see an immediate drop as dotnet GC still has to happen, using `GC.GetTotalMemory(true)` will force GC and you'll see the immediate drop.

Looking as where memory increased after the the clear then there's a memory increase for the PnPContext, this is expected as the last used batch request (the one returning the last 500 list items in this case) is still in memory. When issuing a new batch request the previous executed ones are dropped, so not expecting a continued memory increase here:

![image](https://user-images.githubusercontent.com/7451219/133398359-1369af6e-7a61-49c8-83f9-e54c3dd3ed2c.png)

**Comment by jansenbe** [comment]
@DaleyKD : implementing something similar to Entity Framework's "no tracking" (see https://docs.microsoft.com/en-us/ef/core/querying/tracking#no-tracking-queries) would also help here. That however is big work item, probably something for 2.x series. (I'll create a dedicated issue to track that need)

Are you good with the above feedback?

**Comment by DaleyKD** [comment]
While I'm definitely not happy with the results (not your fault), I think this is accurate.  I am still a bit shocked at how much memory loading a brand new Document Library with 1000 items uses.

Sadly, since we absolutely need to get ALL of the properties (property bag), I can't use View Fields.

I had to redesign my logic to account for this, and it will sadly have a big impact to overall performance.  However, it is what it is.

**Comment by jansenbe** [comment]
@DaleyKD : to accurately get the memory consumption you would need to use a profiling tool, the used counter does not make it clear where the memory consumption actually sits. I've had a look at trimming down the memory footprint and made a change that  does not store all the __deferred  links coming back in a response (although that has no impact for LoadListDataAsStream) which might help. There also have been performance improvements in the last nightly builds (offline unit testing now takes half of the time!), so would recommend checking out a recent build.

Do you really need to have all rows in memory the whole time?

When measuring consumption via the VS profiler the total consumption (heap) was way lower then what was reported by `GC.GetTotalMemory(false)`, I'll see I can find out more about why's that. My testing seemed to indicate around 12MB for 2500 items which is not that bad, but overall reported memory was indeed higher.

**Comment by jansenbe** [comment]
@DaleyKD : Did manage to further lower the memory needs, but more important did also work on performance. In my tests reading 2500 items is approximately double as fast as it used to be (test was offline, so network responsiveness did not impact the results).

Below is the heap size for 2500 items being read in memory:
![image](https://user-images.githubusercontent.com/7451219/134407964-6faf1bd2-0d22-499d-a136-c7789a0382fb.png)

Given the many perf updates from the last days and the fact that the nightly builds are now also signed release builds just like the official ones I would strongly recommend trying with the next nightly (1.3.61 or higher) and see if that helps.

**Comment by jansenbe** [comment]
Hi @DaleyKD ,
To further trim know memory needs we need a solution as mentioned in #547 , for now I'm going to close this issue but feel free to open a new one (or re-open this one) when after testing you see issues.
